[TEST] improve validation of yaml suites

javanna opened the pull request:
Validation of test sections and suites consists of checking that the proper skip features sections are in place depending on the features used in tests.

The validation logic was previously only performed on do sections included in each test section, and the skip needed to be present in the same test section. What happens often though is that the skip is added to the setup section, or the teardown section.

This commit improves the validation of test suites by validating setup and teardown section first, then looking at each test section while still eventually reading the skip section from setup or teardown.

We are also making SkipSection, SetupSection, TearDownSection, ClientYamlTestSection and ClientYamlTestSuite immutable. Previously it was possible to utilize constants like SetupSection.EMPTY, which were modifiable and affect every other future users by modifiying them. This has been corrected.

Also, validation has been improved to cumulate errors so that all the errors from a suite will be listed at once.

This was originated from #34735 and is a prerequisite to get #34735 merged.

javanna committed:
[TEST] improve validation of yaml suites

Validation of test section and suites consists of checking that the proper skip features sections are in place depending on the features used in tests.

The validation logic was previously only performed on do sections included in each test section, and the skip needed to be present in the same test section. What happens often though is that the skip is added to the setup section, or the teardown section.

This commit improves the validation of test suites by validating setup and teardown section first, then looking at each test section while still eventually reading the skip section from setup or teardown.

We are also making SkipSection, SetupSection, TearDownSection, ClientYamlTestSection and ClientYamlTestSuite immutable. Previously it was possible to utilize constants like SetupSection.EMPTY, which were modifiable and affect every other future users by modifiying them. This has been corrected.

Also, validation has been improved to cumulate errors so that all the errors from a suite will be listed at once.

Relates to #34735

elasticmachine commented:
Pinging @elastic/es-core-infra

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java:
```diff
@@ -184,19 +183,44 @@ public static void closeClient() throws IOException {
+                suites.add(suite);
+                try {
+                    suite.validate();
+                } catch(Exception e) {
```
Could we be more precise here? Would seem that not any exception is a validation exception.

javanna replied:
yes that makes sense

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java:
```diff
@@ -33,34 +33,37 @@
         this.name = name;
-        this.executableSections = new ArrayList<>();
+        this.skipSection = skipSection;
+        this.executableSections = executableSections;
```
I think it would be better to wrap the list with unmodifiable here instead if relying on the caller.

javanna replied:
yes, good idea

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java:
```diff
@@ -93,27 +94,29 @@ public static ClientYamlTestSuite parse(String api, String suiteName, XContentPa
         this.name = name;
+        this.setupSection = setupSection;
+        this.teardownSection = teardownSection;
+        this.testSections = testSections;
```
same here.

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java:
```diff
@@ -132,27 +135,52 @@ public SetupSection getSetupSection() {
+    private static void validateExecutableSections(ClientYamlTestSuite yamlTestSuite, List<ExecutableSection> sections,
+                                                   ClientYamlTestSection testSection,
+                                                   SetupSection setupSection, TeardownSection teardownSection) {
+        for (ExecutableSection section : sections) {
```
suggestion: This might read better with `.stream().filter().map()`

javanna replied:
you mean to have the instanceof check in the filter? I am not sure I follow 100% what you are suggesting besides that.

atorok replied:
Yes, the instance-of in the filter and the cast in the map at a minimum, 
but we could also repeat the stream multiple time and transform the  other conditionals into filters as well. 

Come to think of it, we could map to a string containing the error message of the exception that we now throw and collect it to a list which then gets merged into a single one with all the validations and returned. I think it would clean up the code that collects all validation failures. 
Right now we only throw these exceptions so we can lather collect and chain them.

javanna replied:
I pushed an update, we now cumulate even more errors thanks to this change.

alpar-t reviewed (commented):
Left a few comments. 
Would have been easier to review if some changes like the immutability would have been in a separate PR.

javanna committed:
address comments

javanna committed:
Merge branch 'master' into test/improve_yaml_sections_validation

javanna committed:
unused import

javanna committed:
use stream to collect errors

This way we also cumulate multiple errors coming from the same testsection, and multiple errors coming from the same suite (for instance from setup, and also a test section).

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java:
```diff
@@ -139,39 +142,45 @@ public TeardownSection getTeardownSection() {
+
+        Stream<String> errors = sections.stream().filter(section -> section instanceof DoSection)
+            .map(section -> (DoSection) section)
+            .filter(section -> false == section.getExpectedWarningHeaders().isEmpty()
```
I find it more readable to  call `.filter()` twice instead of using `&&` , but might be just me.

javanna replied:
ok I was debating the same, I don't have a preference myself so I will make this change

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java:
```diff
@@ -139,39 +142,45 @@ public TeardownSection getTeardownSection() {
+        errors = Stream.concat(errors, validateExecutableSections(teardownSection.getDoSections(), null, null, teardownSection));
         for (ClientYamlTestSection testSection : testSections) {
-            validateExecutableSections(this, testSection.getExecutableSections(), testSection, setupSection, teardownSection);
+            errors = Stream.concat(errors,
```
This is perfectly fine as it is. Just for the fun of it, could be written as 
```
testSections.stream()
    .flatMap(testSection -> validateExecutableSections(testSection.getExecutableSections(), testSection, setupSection, teardownSection)) )
```

javanna replied:
good one, thanks :)

alpar-t reviewed (approved):
Thanks @javanna  I think this reads much better now.

javanna committed:
review comments

javanna committed:
unused import

javanna committed:
Merge branch 'master' into test/improve_yaml_sections_validation
